File: packages-logicn/logicn-devtools-benchmarks/benchmarks/fibonacci-recursive/bench.cpp
#include <cstdint>
#include <chrono>
#include <iostream>
#include <cstring>
#include <cstdlib>

uint64_t fib(uint64_t n) {
    if (n <= 1) return n;
    return fib(n - 1) + fib(n - 2);
}

int main(int argc, char* argv[]) {
    int n = 30;
    int its = 100000;
    for (int i = 1; i < argc - 1; i++) {
        if (!strcmp(argv[i], "--n")) n = atoi(argv[i+1]);
        if (!strcmp(argv[i], "--operations") || !strcmp(argv[i], "--iterations")) its = atoi(argv[i+1]);
    }
    // Warmup
    fib(n);
    auto t0 = std::chrono::steady_clock::now();
    uint64_t result = 0;
    for (int i = 0; i < its; i++) { result = fib(n); }
    double ms = std::chrono::duration<double, std::milli>(std::chrono::steady_clock::now() - t0).count();
    std::cout << "{\"runtime\":\"cpp\",\"benchmark\":\"fibonacci-recursive-v1\","
              << "\"n\":" << n << ","
              << "\"result\":" << result << ","
              << "\"iterations\":" << its << ","
              << "\"elapsedMs\":" << ms << ","
              << "\"callsPerSecond\":" << its / (ms / 1000) << "}" << std::endl;
}

File: packages-logicn/logicn-devtools-benchmarks/benchmarks/fibonacci-recursive/bench.rs
use std::time::Instant;
use std::env;
use std::hint::black_box;

fn fib(n: u64) -> u64 {
    if n <= 1 { return n; }
    fib(n - 1) + fib(n - 2)
}

fn main() {
    let mut n: u64 = 30;
    let mut iterations: usize = 200;  // fib(30) in Rust ~0.5ms/call; 200 iters ≈ 100ms
    let args: Vec<String> = env::args().collect();
    let mut i = 1;
    while i + 1 < args.len() {
        match args[i].as_str() {
            "--n"          => n          = args[i+1].parse().unwrap_or(n),
            "--operations" | "--iterations" => iterations = args[i+1].parse().unwrap_or(iterations),
            _ => {}
        }
        i += 2;
    }
    // Warmup (with black_box to prevent constant folding)
    let _ = black_box(fib(black_box(n)));
    let t0 = Instant::now();
    let mut result: u64 = 0;
    for _ in 0..iterations {
        result = black_box(fib(black_box(n)));
    }
    let elapsed = t0.elapsed().as_secs_f64() * 1000.0;
    let rate = if elapsed < 0.001 { 0.0 } else { iterations as f64 / (elapsed / 1000.0) };
    println!(r#"{{"runtime":"rust","benchmark":"fibonacci-recursive-v1","n":{},"result":{},"iterations":{},"elapsedMs":{:.3},"iterationsPerSecond":{:.2}}}"#,
        n, result, iterations, elapsed, rate);
}
